packages/client/lib/types.ts

```diff
@@ -1,15 +1,98 @@
-import multiaddr from 'multiaddr'
+import { EventEmitter } from 'events'
+import type multiaddr from 'multiaddr'
+import type { Block, BlockHeader } from '@ethereumjs/block'
 import type Connection from '../../../node_modules/libp2p-interfaces/dist/src/connection/connection'
 import type { MuxedStream } from '../../../node_modules/libp2p-interfaces/dist/src/stream-muxer/types'
+import { Libp2pPeer, Peer, RlpxPeer } from './net/peer'
+import { Peer as Devp2pRlpxPeer } from '@ethereumjs/devp2p'
+import MockPeer from '../test/integration/mocks/mockpeer'
+import { BN } from '../../util/dist'
+/**
+ * Types for the central event bus, emitted
+ * by different components of the client.
+ */
+export enum Event {
+  CHAIN_UPDATED = 'blockchain:chain:updated',
+  SYNC_EXECUTION_VM_ERROR = 'sync:execution:vm:error',
+  SYNC_FETCHER_FETCHED = 'sync:fetcher:fetched',
+  SYNC_SYNCHRONIZED = 'sync:synchronized',
+  SYNC_ERROR = 'sync:error',
+  SYNC_FETCHER_ERROR = 'sync:fetcher:error',
+  PEER_CONNECTED = 'peer:connected',
+  PEER_DISCONNECTED = 'peer:disconnected',
+  PEER_ERROR = 'peer:error',
+  CLIENT_ERROR = 'client:error',
+  CLIENT_LISTENING = 'client:listening',
+  SERVER_LISTENING = 'server:listening',
+  POOL_PEER_ADDED = 'pool:peer:added',
+  POOL_PEER_REMOVED = 'pool:peer:removed',
+  POOL_PEER_BANNED = 'pool:peer:banned',
+  PROTOCOL_ERROR = 'protocol:error',
+  PROTOCOL_MESSAGE = 'protocol:message',
+}
+export interface EventParams {
+  [Event.CHAIN_UPDATED]: []
+  [Event.SYNC_EXECUTION_VM_ERROR]: [Error]
+  [Event.SYNC_FETCHER_FETCHED]: [Block[] | BlockHeader[]]
+  [Event.SYNC_SYNCHRONIZED]: [BN]
+  [Event.SYNC_ERROR]: [Error]
+  [Event.SYNC_FETCHER_ERROR]: [Error, any, any]
+  [Event.PEER_CONNECTED]: [Libp2pPeer | RlpxPeer | Devp2pRlpxPeer | MockPeer | Peer]
+  [Event.PEER_DISCONNECTED]: [Libp2pPeer | RlpxPeer | Devp2pRlpxPeer | MockPeer | Peer]
+  [Event.PEER_ERROR]: [Error, string]
```

Comment:
Unfortunately we are not on TypeScript 4.x yet where we can have [labeled tuple elements](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements), so this section is looking a bit confusing without context of what each type is supposed to represent (although most can be inferred, I think it would be good to have them spelled out to reduce confusion). Can we add a comment after each one labeling what the types are supposed to represent, and after our TS is upgraded we can inline them into the actual code (which should be great for autocomplete). I can help add them too :)

e.g.

`[Event.SYNC_SYNCHRONIZED]: [BN] // [headBlockNum]`

`[Event.PEER_ERROR]: [Error, string] // [error, peerId]`

then after our TS is `>=v4` we'll be able to do:

`[Event.SYNC_SYNCHRONIZED]: [headBlockNum: BN]`

`[Event.PEER_ERROR]: [error: Error, peerId: string]`

Reply:
I've updated all of them except for the `SYNC_FETCHER_ERROR` one.  I looked that one again and I'm honestly not sure what the second and third parameters are for.  The event is emitted [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/57a142d0b9d5ce6c9ba2390eec07c470939b1931/packages/client/lib/sync/fetcher/fetcher.ts#L249) and it's not obvious to me what those two boolean values tell you.   Any ideas?